### PR TITLE
bpo-35493: Use Process.sentinel instead of sleeping for polling worker status in multiprocessing.Pool

### PR DESCRIPTION
**Pull request by pablogsal**
This is a very simple fix for this problem using `Process.sentinel`. If you want a more sophisticated solution, please advice.

```python

import multiprocessing
import time
CONCURRENCY = 1
NTASK = 100
def noop():
    pass
with multiprocessing.Pool(CONCURRENCY, maxtasksperchild=1) as pool:
    start_time = time.monotonic()
    results = [pool.apply_async(noop, ()) for _ in range(NTASK)]
    for result in results:
        result.get()
    dt = time.monotonic() - start_time
    pool.terminate()
    pool.join()
print("Total: %.1f sec" % dt)
```

Before this PR
```
Total: 10.2 sec
```

After this PR:

```
Total: 0.5 sec
```

https://bugs.python.org/issue35493

### CODE REVIEW

File: Lib/multiprocessing/pool.py [hunk 1]
```diff
@@ -437,7 +443,7 @@ def _handle_workers(pool):
         # is terminated.
         while thread._state == RUN or (pool._cache and thread._state != TERMINATE):
             pool._maintain_pool()
-            time.sleep(0.1)
+            pool._wait_for_updates(timeout=0.2)
```
Comment (pablogsal) [inline Lib/multiprocessing/pool.py]: The timeout is needed for detecting changes in `thread._state`

Comment (vstinner) [reply Lib/multiprocessing/pool.py]: asyncio uses the "self-pipe" pattern to wake up itself when it gets an event from a different thread or when it gets a Unix signal. Would it be possible to use a self-pipe (or something else) to wake up the wait when thread._state changes?

Comment (pablogsal) [reply Lib/multiprocessing/pool.py]: We would need to manage also the case when `pool._cache` is empty.

**Comment by pablogsal** [comment]
@vstinner I have used the `self-pipe` pattern to receive notifications on `pool._cache` and `thread._state` changes.

File: Lib/multiprocessing/pool.py (hunk 1 above)
Comment (vstinner) [reply Lib/multiprocessing/pool.py]: > @vstinner I have used the self-pipe pattern to receive notifications on pool._cache and thread._state changes. I maintained the 0.2 timeout for making sure the old behaviour is maintained if some change is not notified using the self._change_notifier queue (by mistake or because of external reasons).

Why 0.2 and not 0.1 or 1.0? I understand that replacing 0.1 with 0.2 doubles the latency of thread pool. Am I right?

File: Lib/multiprocessing/pool.py [hunk 2]
```diff
@@ -596,7 +622,7 @@ def _help_stuff_finish(inqueue, task_handler, size):
             time.sleep(0)
 
     @classmethod
-    def _terminate_pool(cls, taskqueue, inqueue, outqueue, pool,
+    def _terminate_pool(cls, taskqueue, inqueue, outqueue, pool, pool_notifier,
```
Comment (vstinner) [inline Lib/multiprocessing/pool.py]: i would prefer to reuse the same variable name: pool_notifier => change_notifier.

File: Lib/multiprocessing/pool.py [hunk 3]
```diff
@@ -164,7 +174,8 @@ def __init__(self, processes=None, initializer=None, initargs=(),
         self._ctx = context or get_context()
         self._setup_queues()
         self._taskqueue = queue.SimpleQueue()
-        self._cache = {}
+        self._change_notifier = self._ctx.SimpleQueue()
```
Comment (vstinner) [inline Lib/multiprocessing/pool.py:196]: Please add a comment to explain the purpose of this queue. I understand that it exists to wake up _handle_workers(): when _state value is changed.

File: Lib/multiprocessing/pool.py [hunk 4]
```diff
@@ -845,3 +872,6 @@ def _help_stuff_finish(inqueue, task_handler, size):
             pass
         for i in range(size):
             inqueue.put(None)
+
+    def _wait_for_updates(self, timeout):
+        time.sleep(timeout)
```
Comment (vstinner) [inline Lib/multiprocessing/pool.py:949]: I'm now curious: would it be possible to use something smarter to avoid a sleep?

File: Misc/NEWS.d/next/Library/2019-01-09-23-43-08.bpo-35493.kEcRGE.rst [hunk 5]
```diff
@@ -0,0 +1,3 @@
+Use :func:`multiprocessing.connection.wait` instead of polling each 0.2
```
Comment (vstinner) [inline Misc/NEWS.d/next/Library/2019-01-09-23-43-08.bpo-35493.kEcRGE.rst:1]: Would it be possible to describe the higher level effect: reduce the latency on pool operations. I don't recall which operation is impacted. Pool.join()? Or also Pool.terminate()?

Moreover, the change only affects process pool, not thread pool, no?

File: Lib/multiprocessing/pool.py [hunk 6]
```diff
@@ -429,15 +440,28 @@ def _map_async(self, func, iterable, mapper, chunksize=None, callback=None,
...
 
+    def _wait_for_updates(self, timeout):
+        task_queue_sentinels = [self._outqueue._reader.fileno(),
+                                self._outqueue._writer.fileno()]
+        self_notifier_sentinels = [self._change_notifier._reader.fileno()]
+        worker_sentinels = [worker.sentinel for worker in self._pool]
+        sentinels = [*task_queue_sentinels,
+                     *worker_sentinels,
+                     *self_notifier_sentinels]
+        wait(sentinels, timeout=timeout)
```
Comment (vstinner) [inline Lib/multiprocessing/pool.py:499]: Please add a comment to explain why you kept a timeout here.

I dislike keeping the timeout ("active" polling). I would prefer to experiment without timeout. If something hangs: try to understand why and maybe try to fix it. I would only add back the timeout if there is a design issue that we cannot work around or fix.

**Review by vstinner — CHANGES_REQUESTED** [review]

File: Lib/multiprocessing/pool.py (hunk 6 above)
Comment (pablogsal) [reply Lib/multiprocessing/pool.py:499]: If you remove the timeout, everything works as expected, but I prefer to maintain the timeout here for backwards compatibility, so if we forgot something, there is not a deadlock or a hang. The pool has a lot of moving pieces and even if we think we have everything covered we could miss something, so I feel safer to leave the fallback.

The expected behaviour is that the timeout will never be really needed and the underlying select in wait will wake up earlier. Of course, if there are no new jobs/changes/...etc the timeout will be hit but it will do nothing.

What do you think? Do you agree or you really want to delete the timeout? :)

File: Misc/NEWS.d/next/Library/2019-01-09-23-43-08.bpo-35493.kEcRGE.rst (hunk 5 above)
Comment (pablogsal) [reply Misc/NEWS.d/next/Library/2019-01-09-23-43-08.bpo-35493.kEcRGE.rst:1]: > Moreover, the change only affects the process pool, not the thread pool, no?

 Yup, only the process pool.

File: Lib/multiprocessing/pool.py (hunk 4 above)
Comment (pablogsal) [reply Lib/multiprocessing/pool.py:949]: We would need a pipe-based queue (or a raw pipe) to notify a select-based solution. The queues used by the `ThreadPool` are `queue.SimpleQueue()` and are not file-descriptor based so we cannot wait on them on `select`.

We could refactor the general queue to allow a custom notifier in every change but I prefer to do that in another PR.

File: Lib/multiprocessing/pool.py [hunk 7]
```diff
@@ -145,6 +146,15 @@ def _helper_reraises_exception(ex):
 # Class representing a process pool
 #
 
+class PoolCache(dict):
```
Comment (pitrou) [inline Lib/multiprocessing/pool.py]: Can you add a docstring?

File: Lib/multiprocessing/pool.py (hunk 1 above)
Comment (pablogsal) [reply Lib/multiprocessing/pool.py]: I used `0.2` to maintain backwards compatibility with the old behaviour. See my other comment explaining why I maintained the timeout.

File: Lib/multiprocessing/pool.py (hunk 6 above)
Comment (pitrou) [reply Lib/multiprocessing/pool.py:499]: Let's remove the timeout (or make it very long, e.g. 10 seconds). This is only going in 3.8 in any case.

File: Lib/multiprocessing/pool.py [hunk 8]
```diff
@@ -429,15 +440,28 @@ def _map_async(self, func, iterable, mapper, chunksize=None, callback=None,
...
+                     *self_notifier_sentinels]
+        wait(sentinels, timeout=timeout)
+
     @staticmethod
     def _handle_workers(pool):
         thread = threading.current_thread()
 
         # Keep maintaining workers until the cache gets drained, unless the pool
         # is terminated.
-        while thread._state == RUN or (pool._cache and thread._state != TERMINATE):
```
Comment (pitrou) [inline Lib/multiprocessing/pool.py]: Why did you change this?

File: Lib/multiprocessing/pool.py (hunk 4 above)
Comment (pitrou) [reply Lib/multiprocessing/pool.py:949]: Let's keep focused and concentrate on `Pool`. The `ThreadPool` isn't very much used.

**Review by pitrou — CHANGES_REQUESTED** [review]

File: Lib/multiprocessing/pool.py (hunk 8 above)
Comment (pablogsal) [reply Lib/multiprocessing/pool.py]: ~~@pitrou Because we now have notifiers for these events and I want to uncouple the loop with the checks so we could move them in the inner functions if needed. I think the break conditions are a bit more clear now, also.~~

~~Being said that, I am happy to change it back if you prefer, thought.~~

Comment (pablogsal) [reply Lib/multiprocessing/pool.py]: Hummmm, Sorry, I think I am missing a commit here. I will restore this.

File: Lib/multiprocessing/pool.py [hunk 9]
```diff
@@ -429,6 +449,16 @@ def _map_async(self, func, iterable, mapper, chunksize=None, callback=None,
...
 
+    def _wait_for_updates(self, timeout=None):
+        task_queue_sentinels = [self._outqueue._reader.fileno(),
+                                self._outqueue._writer.fileno()]
+        self_notifier_sentinels = [self._change_notifier._reader.fileno()]
+        worker_sentinels = [worker.sentinel for worker in self._pool]
+        sentinels = [*task_queue_sentinels,
+                     *worker_sentinels,
+                     *self_notifier_sentinels]
+        wait(sentinels, timeout=timeout)
```
Comment (vstinner) [inline Lib/multiprocessing/pool.py:499]: Don't you have to consome the _change_notifier queue somewhere, to avoid to get a full queue or just to avoid wasting memory? Call non-blocking get() until it raises an empty error? I suggest to do that after the wait().

File: Lib/multiprocessing/pool.py [hunk 10]
```diff
@@ -429,6 +450,17 @@ def _map_async(self, func, iterable, mapper, chunksize=None, callback=None,
...
+    def _wait_for_updates(self, timeout=None):
+        task_queue_sentinels = [self._outqueue._reader]
+        self_notifier_sentinels = [self._change_notifier._reader]
+        worker_sentinels = [worker.sentinel for worker in self._pool]
+        sentinels = [*task_queue_sentinels,
+                     *worker_sentinels,
+                     *self_notifier_sentinels]
+        wait(sentinels, timeout=timeout)
+        while not self._change_notifier.empty():
+            self._change_notifier.get()
```
Comment (vstinner) [inline Lib/multiprocessing/pool.py]: I don't think that this code is safe, it looks like a race condition: https://en.wikipedia.org/wiki/Time_of_check_to_time_of_use

I suggest to call get(block=False) in a loop until you get an Empty exception.

Note: the race condition is not really critical, since it's fine if we miss a few events.

File: Lib/multiprocessing/pool.py [hunk 11]
```diff
@@ -429,6 +450,17 @@ def _map_async(self, func, iterable, mapper, chunksize=None, callback=None,
...
         return result
 
+    def _wait_for_updates(self, timeout=None):
+        task_queue_sentinels = [self._outqueue._reader]
+        self_notifier_sentinels = [self._change_notifier._reader]
+        worker_sentinels = [worker.sentinel for worker in self._pool]
+        sentinels = [*task_queue_sentinels,
+                     *worker_sentinels,
+                     *self_notifier_sentinels]
+        wait(sentinels, timeout=timeout)
```
Comment (vstinner) [inline Lib/multiprocessing/pool.py:499]: Can you please add a comment on wait() to explain that it completes when at least one sentinel is set and that it's important to not wait until all sentinels completed, but exit frequently to refresh the pool.

This point is non-trivial and it surprised me when I wrote PR #11136, my comment https://github.com/python/cpython/pull/11136#issuecomment-447291966:
> My change doesn't work: self._worker_state_event isn't set when a worker completes, whereas _maintain_pool() should be called frequently to check when a worker completed.

File: Lib/multiprocessing/pool.py (hunk 4 above)
Comment (vstinner) [reply Lib/multiprocessing/pool.py:949]: > Let's keep focused and concentrate on Pool. The ThreadPool isn't very much used.

Sure.

File: Misc/NEWS.d/next/Library/2019-01-09-23-43-08.bpo-35493.kEcRGE.rst (hunk 5 above)
Comment (vstinner) [inline Misc/NEWS.d/next/Library/2019-01-09-23-43-08.bpo-35493.kEcRGE.rst:1]: Oops, I clicked on the wrong button :-( Your NEWS entry still doesn't explain that only process pools are affected.

Another issue: "polling each 0.2 seconds": currently the code uses 0.1 seconds.

**Review by vstinner — COMMENTED** [review]
Oh wow, any multiprocessing is so complex... I'm happy with @pablogsal handles it instead of me :-D

File: Lib/multiprocessing/pool.py (hunk 10 above)
Comment (pitrou) [reply Lib/multiprocessing/pool.py]: There is no race condition as long as this is the only thread that pops from the queue.

File: Lib/multiprocessing/pool.py (hunk 11 above)
Comment (pitrou) [reply Lib/multiprocessing/pool.py:499]: `wait` is already documented: https://docs.python.org/3/library/multiprocessing.html#multiprocessing.connection.wait

Comment (vstinner) [reply Lib/multiprocessing/pool.py:499]: I don't ask to documen the behavior of wait, but more explicit that we stop as soon as the first event complete on purpose.

Comment (pitrou) [reply Lib/multiprocessing/pool.py:499]: It looks obvious to me, especially as the function is named `wait_for_updates`, but I guess it doesn't hurt to add a comment.

File: Lib/multiprocessing/pool.py (hunk 10 above)
Comment (pablogsal) [reply Lib/multiprocessing/pool.py]: Also, sadly `queue.SimpleQueue` has no `block=False` option (we could add one, but I think is not needed).

File: Misc/NEWS.d/next/Library/2019-01-09-23-43-08.bpo-35493.kEcRGE.rst (hunk 5 above)
Comment (pablogsal) [reply Misc/NEWS.d/next/Library/2019-01-09-23-43-08.bpo-35493.kEcRGE.rst:1]: Sorry, I completely forgot about that while fighting Windows issues :/

**Comment by pablogsal** [comment]
I made some changes to eliminate some problems that I found on Windows. I have run all multiprocessing test in a loop with this patch manually on almost all our Windows buildbots and it passes without problems, so I think the pipe/socket solution is resilient also on Windows.

**Comment by pablogsal** [comment]
@vstinner @pitrou I think I have addressed all the review comments. Could you review again? Thanks!

**Comment by pablogsal** [comment]
@vstinner @pitrou I had to rebase since the changes in https://bugs.python.org/issue35378 make the `handle_workers` independent on the pool itself. Some new changes are needed, for example, gathering the sentinels needs to be done in the constructor to avoid references to `self`:

https://github.com/python/cpython/pull/11488/files#diff-2d95253d6de7bbeebbeb131c5f3aecd9R213

Also, the minimal timeout needs to still exist because now that the thread does not keep the pool alive, it needs a wait to exit from `_wait_for_updates` if the pool does not unblock the sentinels because is dead:

https://github.com/python/cpython/pull/11488/files#diff-2d95253d6de7bbeebbeb131c5f3aecd9R493

**Comment by pablogsal** [comment]
Ok, there is a test failure on Windows:

```
test_pool_worker_lifetime (test.test_multiprocessing_spawn.WithProcessesTestPoolWorkerLifetime) ... Exception in thread Thread-908:
Traceback (most recent call last):
  File "C:\projects\cpython\lib\threading.py", line 917, in _bootstrap_inner
    self.run()
  File "C:\projects\cpython\lib\threading.py", line 865, in run
    self._target(*self._args, **self._kwargs)
  File "C:\projects\cpython\lib\multiprocessing\pool.py", line 511, in _handle_workers
    cls._wait_for_updates(sentinels, change_notifier)
  File "C:\projects\cpython\lib\multiprocessing\pool.py", line 494, in _wait_for_updates
    wait(sentinels, timeout=timeout)
  File "C:\projects\cpython\lib\multiprocessing\connection.py", line 878, in wait
    ready_handles = _exhaustive_wait(waithandle_to_obj.keys(), timeout)
  File "C:\projects\cpython\lib\multiprocessing\connection.py", line 810, in _exhaustive_wait
    res = _winapi.WaitForMultipleObjects(L, False, timeout)
OSError: [WinError 6] The handle is invalid
```

File: Lib/multiprocessing/pool.py [hunk 12]
```diff
@@ -189,12 +210,23 @@ def __init__(self, processes=None, initializer=None, initargs=(),
...
+        if hasattr(self._outqueue, "_reader"):
+            task_queue_sentinels = [self._outqueue._reader]
+        else:
+            task_queue_sentinels = []
+        if hasattr(self._change_notifier, "_reader"):
+            self_notifier_sentinels = [self._change_notifier._reader]
+        else:
+            self_notifier_sentinels = []
+
+        sentinels = [*task_queue_sentinels, *self_notifier_sentinels]
```
Comment (pitrou) [inline Lib/multiprocessing/pool.py]: How about a dedicated method that returns this list and that you can override in `ThreadPool`?

File: Lib/multiprocessing/pool.py [hunk 13]
```diff
@@ -452,18 +486,30 @@ def _map_async(self, func, iterable, mapper, chunksize=None, callback=None,
...
-            Pool._maintain_pool(ctx, Process, processes, pool, inqueue,
-                                outqueue, initializer, initargs,
-                                maxtasksperchild, wrap_exception)
-            time.sleep(0.1)
+            cls._maintain_pool(ctx, Process, processes, pool, inqueue,
+                               outqueue, initializer, initargs,
+                               maxtasksperchild, wrap_exception)
+
+            worker_sentinels = [worker.sentinel for worker in
+                                pool if hasattr(worker, "sentinel")]
```
Comment (pitrou) [inline Lib/multiprocessing/pool.py]: Same here: how about a dedicated method to get worker sentinels that you can override in `ThreadPool`?

File: Lib/multiprocessing/pool.py [hunk 14]
```diff
@@ -165,7 +182,11 @@ def __init__(self, processes=None, initializer=None, initargs=(),
         self._ctx = context or get_context()
         self._setup_queues()
         self._taskqueue = queue.SimpleQueue()
-        self._cache = {}
+        # The _change_notifier queue exist to wake ip self._handle_workers()
```
Comment (pitrou) [inline Lib/multiprocessing/pool.py]: "wake up"

File: Lib/multiprocessing/pool.py [hunk 15]
```diff
@@ -165,7 +182,11 @@ def __init__(self, processes=None, initializer=None, initargs=(),
...
+        # when the cache (self._cache) is empty or when ther is a change in
```
Comment (pitrou) [inline Lib/multiprocessing/pool.py]: "when there is"

File: Lib/multiprocessing/pool.py [hunk 16]
```diff
@@ -145,6 +147,21 @@ def _helper_reraises_exception(ex):
...
+    notification is done by the use of a queue that is provided when
+    instantiating the cache.
+    """
+    def __init__(self, *args, notifier=None, **kwds):
+        self.notifier = notifier
+        super().__init__(*args, **kwds)
+    def __delitem__(self, item):
+        super().__delitem__(item)
+        if not self:
+            self.notifier.put(None)
```
Comment (pitrou) [inline Lib/multiprocessing/pool.py:171]: Can you add a comment explaining why it's important to wake up when the cache is emptied?

**Review by pitrou — CHANGES_REQUESTED** [review]
A few minor points below.

File: Lib/multiprocessing/pool.py (hunk 13 above)
Comment (pablogsal) [reply Lib/multiprocessing/pool.py]: The problem here is that we don't have the reference to the pool object (`self`) after https://bugs.python.org/issue35378 so we can only call class/static methods. The most we can do here is make a static/class method that takes the list of workers (called `pool` here) and then it returns the list of sentinels.

What do you think?

Comment (pitrou) [reply Lib/multiprocessing/pool.py]: That sounds fine to me.

File: Lib/multiprocessing/pool.py [hunk 17]
```diff
@@ -233,6 +265,8 @@ def __del__(self, _warn=warnings.warn, RUN=RUN):
         if self._state == RUN:
             _warn(f"unclosed running multiprocessing pool {self!r}",
                   ResourceWarning, source=self)
+            if getattr(self, '_change_notifier') is not None:
```
Comment (vstinner) [inline Lib/multiprocessing/pool.py]: Why not just self._change_notifier?

getattr(obj, attr) raises AttributeError if the attribute doesn't exist. Maybe you want to write getattr(self, '_change_notifier', None)?

File: Lib/multiprocessing/pool.py [hunk 18]
```diff
@@ -145,6 +147,21 @@ def _helper_reraises_exception(ex):
...
+class _PoolCache(dict):
+    """
+    Class that implements a cache for the Pool class that will notify
+    the pool management threads every time the cache is emptied. The
+    notification is done by the use of a queue that is provided when
+    instantiating the cache.
+    """
+    def __init__(self, *args, notifier=None, **kwds):
+        self.notifier = notifier
+        super().__init__(*args, **kwds)
```
Comment (vstinner) [inline Lib/multiprocessing/pool.py:159]: PEP 8, please add an empty line between methods.

**Comment by vstinner** [comment]
I tested manually example from https://bugs.python.org/issue35493#msg331797:

* Without the change: 10.2 sec
* With the change: 0.3 sec

It's 34x faster, nice :-)

**Review by vstinner — APPROVED** [review]
LGTM, but I would prefer that @pitrou or @applio also review the change.

LGTM means that I reviewed the change and it seems like you covered any changes which can wake up _wait_for_updates().

It seems like @pitrou proposed a different implementation, but I don't recall care of this level of detail. I let you deal with that :-)

I tested manually that the PR fix the bug that I reported: see my previous comment and https://bugs.python.org/issue35493#msg331797 initial message.

**Comment by pablogsal** [comment]
@pitrou  Commit 41cf470 should address everything. Could you take a final look?

File: Lib/multiprocessing/pool.py [hunk 19]
```diff
@@ -233,13 +264,25 @@ def __del__(self, _warn=warnings.warn, RUN=RUN):
...
 
+    def _get_sentinels(self):
+        task_queue_sentinels = [self._outqueue._reader]
+        self_notifier_sentinels = [self._change_notifier._reader]
+        return [*task_queue_sentinels, *self_notifier_sentinels]
+
+    @staticmethod
+    def _get_worker_sentinels(workers):
+        return [worker.sentinel for worker in
+                workers if hasattr(worker, "sentinel")]
```
Comment (pitrou) [inline Lib/multiprocessing/pool.py:284]: Is the condition still useful?

**Review by pitrou — APPROVED** [review]

**Comment by pitrou** [comment]
I added a minor comment, otherwise LGTM. Thanks @pablogsal !

File: Lib/multiprocessing/pool.py (hunk 19 above)
Comment (pablogsal) [reply Lib/multiprocessing/pool.py:284]: Yup, `DummyProcess` does not have a "sentinel" attribute.

**Comment by pablogsal** [comment]
Hummm...weird. GitHub has notified me that there was a problem merging the PR and when I clicked "retry" it has not used the message I wrote for the commit.

Anyway, thank you everyone that participated in the review :)

**Comment by rgommers** [comment]
@pablogsal this seems to have caused https://bugs.python.org/issue38501 (hangs on both macOS and Windows with Python 3.8). That bug shows up in multiple SciPy modules as hangs (e.g. https://github.com/scipy/scipy/issues/11835). Could you please have a look at `bpo-38501`?
